csharp/src/SeedLang.Runtime/EventHandler.cs

```diff
@@ -0,0 +1,42 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+
+namespace SeedLang.Runtime {
+  // The interface of subscribable objects.
+  internal interface ISubscribable<in Visualizer> {
+    void Subscribe(Visualizer visualizer);
+    void Unsubscribe(Visualizer visualizer);
+  }
+
+  // The event handler to notify all the subscribed visualizers when the event is triggered.
+  public class EventHandler<Event> : ISubscribable<IVisualizer<Event>> {
+    private readonly List<IVisualizer<Event>> _visualizers = new List<IVisualizer<Event>>();
+
+    public void Subscribe(IVisualizer<Event> visualizer) {
```

Comment:
The names look confusing to me.

> **Subscribe**: arrange to receive something, typically a publication, regularly by paying in advance.
"subscribe to the magazine for twelve months and receive a free limited-edition T-shirt"

In our context, it's okay to say "an event receiver (or event handler) subscribes to an event type published by an event publisher", where:

* Event receiver/handler -> a concrete visualizer
* Event type -> event type
* Event publisher -> the class that sends the event to the visualizer

A couple of confusing things here:

* You named this class "EventHander". This class is actually sending events to the registered visualizers. So this class is just like an event publisher, while the concrete visualizers are more like "event handlers" than this class.
* "Subscribe" is mostly used in the context that "a consumer subscribes to a message type". Here, we actually want "a visualizer subscribes to an event type". But the current code reads "an event publisher subscribes to a visualizer(event handler)".

See [Kafka's subscribe API](https://kafka.apache.org/28/javadoc/org/apache/kafka/clients/consumer/Consumer.html#subscribe(java.util.Collection) for a reference to define a subscribe method from a consumer side:

```
Consumer.subscribe
void subscribe​(Collection<String> topics)
```

From a publisher side, similar functionalities can be defined as:

```
Publisher.addEventListener(...);
Publisher.removeEventListener(...);
```

Reply:
I refer to the naming of Microsoft observer design pattern: [https://docs.microsoft.com/en-us/dotnet/standard/events/observer-design-pattern](https://docs.microsoft.com/en-us/dotnet/standard/events/observer-design-pattern). Maybe different person has dofference interpreter of this naming.
But I think your suggestion about publisher might be better. I will try to change the naming accordingly.

Reply:
I guess Microsoft doesn't have a good reputation in the API design area. Considering the following code:

The example code at https://docs.microsoft.com/en-us/dotnet/api/system.iobservable-1?view=net-5.0

```
public class LocationReporter : IObserver<Location>
{
  public virtual void Subscribe(IObservable<Location> provider)
   {
      if (provider != null)
         unsubscriber = provider.Subscribe(this);
   }
}
```

See? Microsoft uses the name "Subscribe" twice, with totally different meanings :-) In Microsoft's logic, an observer can "subscribe" to an observable object, and, on the other hand, an observable object (provider) can also "subscribe" to an observer. Super confusing!

In _Design Patterns_ there is a subscriber example under the pattern **Observer**. See "Example - News Agency at "https://www.oodesign.com/observer-pattern.html. The registration action is initiated by the observable object (or the publisher), but the methods name are "attach" and "detach". This name pair is way better than Microsoft's solution.

So, I think there are a couple of good practices we can follow:

A) Message queue like pub-sub pattern:

```
Publisher.register(eventTypeOrTopic)
Publisher.unregister(eventTypeOrTopic)
Consumer.subscribe(publisher, eventTypeOrTopic)
Consumer.unsubscribe(publisher, eventTypeOrTopic)
```

B) JavaScript like event-listener pattern:

```
Element.dispatchEvent(eventType);
Element.addEventListener(eventHandler);
Element.removeEventListener(eventHandler);
```

C) _Design Patterns_'s observer pattern:

```
Publisher.attach(subscriber);
Publisher.detach(subscriber);
Publisher.notifyObservers();
Subscriber.update();
```

If our Visualization mechanism will be used in both sync and async circumstances, I'd equally prefer choice A) or choice B). Otherwise, if we only consider the sync case, I'd prefer choice C).

Reply:
Choose A)